### ISSUE TITLE
#shooting failed error

### ISSUE THREAD

**Issue by shan1525** [brinckmann/montepython_public#184]
Hi everyone ,

I am trying to run a modified CLASS with montepython with a scalar field approach . Everything went quite smoothly as follows:

#-----CLASS v2.8.2 (branch: , hash: )-----

#------Experiments to test (separated with commas)-----

data.experiments=['Planck_highl_TT', 'Planck_lowl_EE', 'Planck_lowl_TT', 'Planck_lensing']

# Cosmological parameters list

data.parameters['omega_b']      = [  2.2377,   None, None,      0.015, 0.01, 'cosmo']
data.parameters['omega_cdm']    = [ 0.12010,   None, None,     0.0013,    1, 'cosmo']
data.parameters['100*theta_s']  = [ 1.04110,   None, None,    0.00030,    1, 'cosmo']
data.parameters['ln10^{10}A_s'] = [  3.0447,   None, None,      0.015,    1, 'cosmo']
data.parameters['n_s']          = [  0.9659,   None, None,     0.0042,    1, 'cosmo']
data.parameters['scf_parameters__1']        = [  3.0,      None, None,          0,    1, 'cosmo']
data.parameters['scf_parameters__2']        = [  3.0,     1.0, 9.0,        0.5,    1e26, 'cosmo']
data.parameters['scf_parameters__3']        = [  3.0,     1.0, 9.0,        0.5,    1e-26, 'cosmo']
data.parameters['scf_parameters__4']        = [  1.0,      None, None,          0,    1, 'cosmo']
data.parameters['scf_parameters__5']        = [  2.6,       0.9, 3.0,       0.05,    1, 'cosmo']
data.parameters['scf_parameters__6']        = [  0.0,      None, None,          0,    1, 'cosmo']

# Derived parameters

data.parameters['z_reio']          = [1, None, None, 0,     1,   'derived']
data.parameters['YHe']             = [1, None, None, 0,     1,   'derived']
data.parameters['H0']              = [0, None, None, 0,     1,   'derived']
data.parameters['A_s']             = [0, None, None, 0,  1e-9,   'derived']
data.parameters['sigma8']          = [0, None, None, 0,     1,   'derived']
data.parameters['log10z_c']        = [1, None, None, 0,     1,   'derived']
data.parameters['fEDE']            = [0, None, None, 0,     1,   'derived']

# For Scalar Field Dynamics
data.cosmo_arguments['Omega_Lambda'] = 0
data.cosmo_arguments['Omega_fld'] = 0
data.cosmo_arguments['Omega_scf'] = -1
data.cosmo_arguments['attractor_ic_scf']   = 'no'
data.cosmo_arguments['scf_tuning_index']   = 3

#-----Likelihood-Planck_lowl_EE-----
Planck_lowl_EE.path_clik = '/home/shanno/Desktop/monte/code/plc_3.0/plc-3.01/../../../baseline/plc_3.0/low_l/simall/simall_100x143_offlike5_EE_Aplanck_B.clik'

#-----Likelihood-Planck_lowl_TT-----
Planck_lowl_TT.path_clik = '/home/shanno/Desktop/monte/code/plc_3.0/plc-3.01/../../../baseline/plc_3.0/low_l/commander/commander_dx12_v3_2_29.clik'

#-----Likelihood-Planck_lensing-----
Planck_lensing.path_clik = '/home/shanno/Desktop/monte/code/plc_3.0/plc-3.01/../../../baseline/plc_3.0/lensing/smicadx12_Dec5_ftl_mv2_ndclpp_p_teb_consext8.clik_lensing'

#-----------Cosmological-arguments---------
data.cosmo_arguments.update({'sBBN file': '/home/shanno/Desktop/monte/class_ede-master/class/bbn/sBBN.dat', 'Omega_scf': -1, 'P_k_max_h/Mpc': 1.0, 'output': 'mPk tCl lCl pCl ', 'N_ur': 2.0328, 'Omega_Lambda': 0, 'attractor_ic_scf': 'no', 'Omega_fld': 0, 'scf_tuning_index': 3, 'm_ncdm': 0.06, 'l_max_scalars': 2508, 'k_pivot': 0.05, 'lensing': ' yes ', 'non linear': 'halofit', 'N_ncdm': 1, 'T_ncdm': 0.71611})

But then it is giving the following error and montepython run is stopped:

Error in Class: background_init(L:693) :condition (pba->shooting_failed == _TRUE_) is true; Shooting failed, try optimising input_get_guess(). Error message:

input_init(L:436) :error in fzero_Newton(input_try_unknown_parameters, x_inout, dxdF, unknown_parameters_size, 1e-4, 1e-6, &fzw, &fevals, errmsg);
=>fzero_Newton(L:1115) :error in func(x_inout, x_size, param, F0, error_message);
=>input_try_unknown_parameters(L:3584) :error in background_init(&pr,&ba);
=>background_init(L:736) :error in background_solve(ppr,pba);
=>background_solve(L:1735) :error in generic_integrator(background_derivs, tau_start, tau_end, pvecback_integration, &bpaw, ppr->tol_background_integration, ppr->smallest_allowed_variation, &gi);
=>generic_integrator(L:112) :error in rkqs(&x, h, eps, &hdid, &hnext, derivs, parameters_and_workspace_for_derivs, pgi);
=>rkqs(L:156) :condition (xnew == *x) is true; stepsize underflow at x=4.634896e-09
Traceback (most recent call last):
  File "montepython/MontePython.py", line 40, in <module>
    sys.exit(run())
  File "/home/shanno/Desktop/monte/montepython_public-3.3/montepython/run.py", line 45, in run
    sampler.run(cosmo, data, command_line)
  File "/home/shanno/Desktop/monte/montepython_public-3.3/montepython/sampler.py", line 46, in run
    mcmc.chain(cosmo, data, command_line)
  File "/home/shanno/Desktop/monte/montepython_public-3.3/montepython/mcmc.py", line 787, in chain
    newloglike = sampler.compute_lkl(cosmo, data)
  File "/home/shanno/Desktop/monte/montepython_public-3.3/montepython/sampler.py", line 776, in compute_lkl
    value = likelihood.loglkl(cosmo, data)
  File "/home/shanno/Desktop/monte/montepython_public-3.3/montepython/likelihood_class.py", line 953, in loglkl
    cl = self.get_cl(cosmo)
  File "/home/shanno/Desktop/monte/montepython_public-3.3/montepython/likelihood_class.py", line 192, in get_cl
    cl = cosmo.lensed_cl(int(l_max))
  File "classy.pyx", line 567, in classy.Class.lensed_cl
classy.CosmoSevereError: 

Error in Class: No lensed Cl computed.....

Can somebody suggest a solution. I am stuck at this point. As I am very new to this python module , I could not find any solution by myself. So , If anybody has an answer please suggest. Hi @brinckmann  can you please look into this and suggest some idea.

regards,
shan

**Comment by brinckmann** [comment]
Hi Shan,

This appears to be a CLASS problem, rather than a MontePython one. If you want to solve the problem you should post on that github page instead.

That said, if it happens infrequently you might be able to simply reject the sampled points in an MCMC chain without biasing the results, but it would have to be sufficiently rare. You could do this in two ways, either change the type of the error on the CLASS side to something less severe (so MontePython doesn't abort) or you can catch the error on the MontePython side and return a very large loglkl value so the point is rejected by the sampler. 

The latter might be more risky, as it might catch other severe errors and ignore them. At the very least I would suggest outputting the error message so you can see if anything potentially problematic happened. What you want to keep an eye out for is if there's a systematic bias in parameters rejected due to errors, e.g. all of them have low or high values of some parameter. If the rejections are infrequent or appear random it _might_ be OK to skip them.

The lines in question are line 758-761 of `montepython/sampler.py` . As you can see it currently raises an error if a `CosmoSevereError` is encountered. If you instead have `return data.boundary_loglike` like on the preceding line the code would continue, with the big caveat that it would ignore _all_ severe errors from CLASS. To output the error do something like on line 755-756. Again, I will reiterate that suppressing all CosmoSevereError may lead to problems for you down the road, so think about whether you can solve this in a different way.

Best,
Thejs

**Comment by shan1525** [comment]
Thank you so much for your reply @brinckmann . I will try to follow the leads that you have suggested. At least I got the direction to go to otherwise I would have been wandering to search for the possible solutions. Surely, I will raise this issue in CLASS page and see whether I get any response. If I get any response from there, I will post that in here so that it will be helpful for the people having similar issues. Again thanks for your suggestion.

**Comment by miug14** [comment]
Hi @shan1525, I have raised this issue in the CLASS git page as well so feel free to comment in there as well!

**Comment by shan1525** [comment]
Thanks @miug14  for the help. I will comment  there. And let's hope we get some solutions or suggestions .

**Comment by shan1525** [comment]
Thank you so much  @brinckmann and @miug14 for your good suggestions and co-operations. My problem has been resolved , all thanks to you guys. It was the error of the range of the parameter space. I was giving a 10sigma range , so that I could get the idea of max likelihood, but in doing so , it was going out of the budgeted equation and the CLASS was failing with the shooting error. As soon as I changed the range of the scalar field parameter values around the best fit priors , the Montepython ran smoothly. That's why I wanted to thank you guys for your help otherwise I would have been stuck for a long time. 

Best regards,
shan.
